Add pluck waste edges functions.

seigodev opened the pull request:
間引きを行う関数の追加。
・pluck_waste_edges( ): 間引くエッジを除去する。
・make_node2ancestors_recursively(): keyがノードでvalueがノードの全祖先となる辞書を作る。
・search_removable_edges(): 間引けるエッジを探す。
・make_union_by_dict_value(): 値がセットの辞書から、全値の和集合を作成する。

seigodev committed:
Add pluck waste edges functions.

間引きを行う関数の追加。
・pluck_waste_edges( ): 間引くエッジを除去する。
・make_node2ancestors_recursively(): keyがノードでvalueがノードの全祖先となる辞書を作る。
・search_removable_edges(): 間引けるエッジを探す。
・make_union_by_dict_value(): 値がセットの辞書から、全値の和集合を作成する。

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+
+    ancestors = set()
+    for target in node.targets:
+        ancestors = ancestors | {target}
```
```python
ancestors |= {target}
```
へ変更. 他も同様

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+        removable_edge_list = search_removable_edges(node, node2ancestors)
+        for source, target in removable_edge_list:
+            source.targets.remove(target)
+            target.sources.remove(source)
```
```python
    for node in nodes:
        make_node2ancestors_recursively(node, node2ancestors)

    for node in nodes:
        removable_edge_list = search_removable_edges(node, node2ancestors)
        for source, target in removable_edge_list:
            source.targets.remove(target)
            target.sources.remove(source)
```
と分ける

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+        nodeにターゲットが存在しない：要素がnodeのみのセット
+        それ以外：nodeの全祖先のノードのセット
+    """
+    if not node.targets:
```
この前に
```python
if node in node2ancestors:
    return node2ancestors[node]
```
を入れる

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+        if target in node2ancestors.keys():
+            ancestors = ancestors | node2ancestors[target]
+        else:
+            ancestors = ancestors | make_node2ancestors_recursively(target, node2ancestors)
```
前のif文は関数の冒頭で調べさせるので不要．
```python
ancestors |= make_node2ancestors_recursively(target, node2ancestors)
```
でよい．

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+        node: 間引きたいノード(ソース側)。
+        node2ancestors: key=nodeのtarget, value=keyの全祖先のノードのセット の辞書。
+    Return:
+        thin_out_edge_list: 間引いてよいエッジ(source, target)のリスト。
```
thin_out_edge_list -> removable_edge_list

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+    # node2ancestorsの中からkeyがnode.targets内に存在するもののみ取ってくる。
+    target2ancestors = \
+        {target: ancestors for target, ancestors in node2ancestors.items() if target in node.targets}
+    target_route_union = make_union_by_dict_value(target2ancestors)
```
target_route_union -> all_target_ancestorsへ名称変更
以下へソースコードを書き換え
```python
all_target_ancestors = set()
for target in node.targets:
    all_target_ancestors |= target2ancestors[target]
```

aabaa commented on demo/create_graph.py:
```diff
@@ -128,6 +128,89 @@ def create_node_list(input_node_dict):
+    set_union = set()
+    for v in set_dict.values():
+        set_union = set_union | v
+    return set_union
```
削除

seigodev committed:
Fix requested changes.

・ancestor = ancestor | {node} -> ancestor |= {node}  (176, 177, 195行)
・pluck_waste_edges() -> remove_redundant_dependency() (136行)
・remove_redundant_dependency()の処理を二つに分ける。(147~150行)
・make_node2ancestors_recursively()の処理の改善。(167~177行)
・search_removable_edges() -> search_removable_dependency() (182行)
・target_route_union -> all_target_ancestors, およびall_target_ancestorsのコード変更(192~199行)

seigodev committed:
Fix main function.

pluck_waste_edges() -> remove_redundant_dependency()
